src/merkle_tree/constraints.rs

```diff
@@ -101,75 +45,251 @@ where
+        // To traverse up a MT, we iterate over the path from bottom to top (in reverse)
+
+        let path_rev: Vec<_> = self.path.iter().rev().collect();
+
+        let auth_path_rev: Vec<_> = self.auth_path.iter().rev().collect();
+
+        for level in 0..auth_path_rev.len() {
+            // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+            // and the bit being 1 indicates our currently hashed value is on the right.
+            // Thus `left_hash` is the sibling if bit is 1, and it's the computed hash if bit is 0
+            let bit = path_rev[level];
+            let sibling = auth_path_rev[level];
+
+            let left_hash = bit.select(sibling, &curr_hash)?;
+            let right_hash = bit.select(&curr_hash, sibling)?;
+
+            curr_hash = TwoToOneH::evaluate_both(
+                two_to_one_hash_parameter,
+                &left_hash.to_bytes()?,
+                &right_hash.to_bytes()?,
+            )?;
+        }
+
+        Ok(curr_hash)
+    }
+
+    /// Given a leaf, calculate the hash of the merkle tree
+    /// along the path and check if the given root is correct.
+    ///
+    /// Constraints will not be satisfied if root is incorrect.
+    pub fn check_membership(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<(), SynthesisError> {
+        self.conditionally_check_membership(
+            leaf_hash_parameter,
+            two_to_one_hash_parameter,
+            root,
+            leaf,
+            &Boolean::constant(true),
+        )
+    }
+
+    /// Check membership if `should_enforce` is True.
+    pub fn conditionally_check_membership(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+        should_enforce: &Boolean<ConstraintF>,
+    ) -> Result<(), SynthesisError> {
+        let expected_root =
+            self.calculate_root(leaf_hash_parameter, two_to_one_hash_parameter, leaf)?;
+        // enforce `curr_hash` is root
+        expected_root.conditional_enforce_equal(root, should_enforce);
+        Ok(())
+    }
+
+    /// update the `old_leaf` to `new_leaf` and returned the updated MT root.
+    ///
+    /// If the `old_leaf` does not lead to `old_root`, constraints will not be satisfied.
+    pub fn update_leaf(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        old_root: &TwoToOneH::OutputVar,
+        old_leaf: &impl ToBytesGadget<ConstraintF>,
+        new_leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<TwoToOneH::OutputVar, SynthesisError> {
+        self.conditionally_check_membership(
```

Comment:
Should this instead be `calculate root`, and than an equality check, since theres no condition here?

Reply:
Actually is this function supposed to check the original leaf? Seems like it shouldn't, and checking and updating should be what `check_and_update` does?

Reply:
Current workflow is:
- Check if `old_root` is correct
- Get the updated path using `new_leaf`
Alternative workflow is
- Get the updated path using `new_leaf`
- Do equality check, checking that the root of new path is correct. 
The current implementation is essentially the current workflow. Order does not matter here since nothing is mutated here. But I believe what you mean is the alternative workflow?

Reply:
If we use alternative workflow instead, I might need to change replace `old_root` with `new_root`, assuming the prover knows the updated root before running this fucntion

Reply:
Sorry, I deleted a comment (which I realized didn't make sense)

I was wondering if we should change this method called to `calculate_root`, and directly checking equality, instead of `conditionally_check_membership`, since we're not actually using it conditionally. 

It doesn't really matter though, since the equality check in both cases creates one extra constraint AFAIU. (Whereas it doesn't actually need to add any constraints in the future w/ compiler optimizations)

Reply:
In fact, alternative workflow is already implemented in `update_and_check` lol (while we also check the correct of old root before getting the updated path).

Reply:
oh lol, I think its fine either way

Reply:
I will use `check_membership` instead for clarification (but those two are essentially equivalent in performance, as `check_membership` calls `conditionally_check_membership(..., condition=True)`).